core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -130,4 +151,81 @@ object Scheduler {
       }
     }
   }
+
+  private def persistenceFlow(
+      podRecordRepository: PodRecordRepository
+  ): Flow[SchedulerEvents, SchedulerEvents, NotUsed] = {
+    Flow.fromGraph(new GraphStage[FlowShape[SchedulerEvents, SchedulerEvents]] {
+
+      private val inlet = Inlet[SchedulerEvents]("scheduler-events-inlet")
+      private val outlet = Outlet[SchedulerEvents]("scheduler-events-outlet")
+
+      override def shape: FlowShape[SchedulerEvents, SchedulerEvents] =
+        new FlowShape[SchedulerEvents, SchedulerEvents](inlet, outlet)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+        new GraphStageLogic(shape) {
+          private[this] var nextElement: Option[SchedulerEvents] = None
+          val eventCounter = new AtomicInteger(0)
+
+          def maybePushToPull(): Unit = {
+            if (nextElement.isDefined && isAvailable(outlet)) {
+              push(outlet, nextElement.get)
+              nextElement = None
+            }
+            if (nextElement.isEmpty && !hasBeenPulled(inlet)) {
+              pull(inlet)
+            }
+          }
+
+          setHandler(
+            inlet,
+            new InHandler {
+              override def onPush(): Unit = {
+                if (nextElement.isEmpty) {
+                  val schedulerEvents = grab(inlet)
+                  val (storeRecords, deleteRecords) = schedulerEvents.stateEvents.collect {
+                    case x: PodRecordUpdated => x
+                  }.partition(_.newRecord.isDefined)
+                  eventCounter.addAndGet(storeRecords.size + deleteRecords.size)
+                  nextElement = Some(schedulerEvents)
+                  if (eventCounter.get() == 0) {
+                    maybePushToPull()
+                  } else {
+                    // TODO : batch transactions
+                    Source(storeRecords)
+                      .map(_.newRecord.get)
```

Comment:
Is `newRecord` an `Option`? If so I'd use `collect`:
```suggestion
                      .collect { case Some(newRecord) => newRecord }
```